library/src/layout/par.rs

```diff
@@ -1170,6 +1194,21 @@ fn line<'a>(
         }
     }
 
+    if start_cjk_punct {
+        let reshaped = first.as_mut().or(last.as_mut()).and_then(Item::text_mut);
+        if let Some(reshaped) = reshaped {
+            if let Some(punct) = reshaped.glyphs.to_mut().first_mut() {
```

Comment:
why does all of this happen always whereas the end thing only happens in the `if let`?

Reply:
Because the first `inner.split_last()` will always return `Some` if the text is not empty. We can always find line end in `shaped` here.

https://github.com/typst/typst/blob/1d8b68bc1373860fbdaeaafaf093736db5d18f3c/library/src/layout/par.rs#L1121-L1123

But in the second split, the line start can be in the `first` variable or in the `inner`. We cannot just do that in the `if let`.

Reply:
Makes sense.